internal/cluster/distribution/eks/eksadapter/node_pool_manager.go

```diff
@@ -123,45 +126,91 @@ func generateNodePoolStackName(clusterName string, poolName string) string {
 }
 
 // ListNodePools lists node pools from a cluster.
-func (n nodePoolManager) ListNodePools(ctx context.Context, cluster cluster.Cluster, nodePoolNames []string) ([]eks.NodePool, error) {
-	clusterClient, err := n.dynamicClientFactory.FromSecret(ctx, cluster.ConfigSecretID.String())
+func (n nodePoolManager) ListNodePools(
+	ctx context.Context,
+	c cluster.Cluster,
+	existingNodePools map[string]eks.ExistingNodePool,
+) (nodePools []eks.NodePool, err error) {
+	if c.ConfigSecretID.ResourceID == "" || // Note: cluster is being created or errorred before k8s secret would be available.
+		c.Status == cluster.Deleting {
+		return nil, cluster.NotReadyError{
+			OrganizationID: c.OrganizationID,
+			ID:             c.ID,
+			Name:           c.Name,
+		}
+	}
+
+	clusterClient, err := n.dynamicClientFactory.FromSecret(ctx, c.ConfigSecretID.String())
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating dynamic Kubernetes client factory failed",
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
+			"configSecretID", c.ConfigSecretID.String(),
+		)
 	}
 
 	manager := npls.NewManager(clusterClient, n.namespace)
 	labelSets, err := manager.GetAll(ctx)
 	if err != nil {
 		return nil, errors.WrapWithDetails(err, "retrieving node pool label sets failed",
-			"cluster", cluster,
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
 			"namespace", n.namespace,
 		)
 	}
 
-	awsClient, err := n.awsFactory.New(cluster.OrganizationID, cluster.SecretID.ResourceID, cluster.Location)
+	awsClient, err := n.awsFactory.New(c.OrganizationID, c.SecretID.ResourceID, c.Location)
 	if err != nil {
-		return nil, errors.WrapWithDetails(err, "creating aws factory failed", "cluster", cluster)
+		return nil, errors.WrapWithDetails(err, "creating aws factory failed",
+			"OrganizationID", c.OrganizationID,
+			"clusterID", c.ID,
+			"clusterName", c.Name,
+		)
 	}
 
 	cfClient := n.cloudFormationFactory.New(awsClient)
 	describeStacksInput := cloudformation.DescribeStacksInput{}
-	nodePools := make([]eks.NodePool, 0, len(nodePoolNames))
-	for _, nodePoolName := range nodePoolNames {
-		stackName := generateNodePoolStackName(cluster.Name, nodePoolName)
-		describeStacksInput.StackName = &stackName
+	nodePools = make([]eks.NodePool, 0, len(nodePools))
+	for _, existingNodePool := range existingNodePools {
+		nodePools = append(nodePools, eks.NodePool{
+			Name:   existingNodePool.Name,
+			Labels: labelSets[existingNodePool.Name],
+		})
+		nodePool := &nodePools[len(nodePools)-1]
```

Comment:
If the appending is moved to the end of the loop, we can avoid looking the nodepool up from the slice. Then it also doesn't have to be a pointer.

Reply:
If we move it there, we need to append at every early continue in "error handling" branches.

Is that preferred over this approach?

Reply:
Hm....

Maybe extract the body of the loop into a separate function and just return with a node pool?

That sounds better anyway than using a lot of continues.

Also, you could even extract some of the logic into smaller functions and return a node pool directly (functional approach).

For example this part:

> err = sdkCloudFormation.ParseStackParameters(stack.Parameters, &nodePoolParameters)

could go into a separate function, in the err case set the status and message and return a node pool, otherwise fill in the values in the node pool struct and return it.

It's up to you, not sure it's worth blocking the PR or rewriting the code, they are functionally the same, but it feels less readable now.

Reply:
I agree, I like this approach, so I'm gonna implement it.

Reply:
[Fixed](https://github.com/banzaicloud/pipeline/pull/3187/commits/548febed09b25fe668a14074e8df06f0040031e9).